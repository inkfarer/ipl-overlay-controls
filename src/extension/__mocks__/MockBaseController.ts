import { MessageInputMap, MessageResultMap } from '../../types/messages/MessageMaps';
import { BaseController } from '../controllers/BaseController';

export let controllerListeners:
    {[key in keyof MessageInputMap]?: (data?: unknown) => MessageResultMap[key] | Promise<MessageResultMap[key]>} = {};

beforeEach(() => {
    jest.spyOn(BaseController.prototype, 'listen').mockImplementation(
        <key extends keyof MessageInputMap>
        (name: key, callback: (data?: unknown) => any) => {
            controllerListeners[name] = callback;
        });

    controllerListeners = {};
});
