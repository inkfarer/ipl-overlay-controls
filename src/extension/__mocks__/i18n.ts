import i18next from 'i18next';

beforeAll(async () => {
    await i18next.init({
        lng: 'cimode',
        appendNamespaceToCIMode: true
    });
});
