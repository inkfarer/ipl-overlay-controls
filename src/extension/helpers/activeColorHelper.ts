import { ColorWithCategory, SetActiveColorRequest } from '../../types/messages/activeRound';
import * as nodecgContext from './nodecg';
import { ActiveRound, RuntimeConfig, SwapColorsInternally } from '../../types/schemas';
import cloneDeep from 'lodash/cloneDeep';
import { ColorGroup } from '../../types/colors';
import { perGameData } from '../../helpers/gameData/gameData';
import { swapColors } from '../../helpers/ColorHelper';

const nodecg = nodecgContext.get();

const activeRound = nodecg.Replicant<ActiveRound>('activeRound');
const runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');
const swapColorsInternally = nodecg.Replicant<SwapColorsInternally>('swapColorsInternally');

export function setActiveColor(data: SetActiveColorRequest): void {
    const newActiveRound = cloneDeep(activeRound.value);
    newActiveRound.activeColor = {
        categoryName: data.categoryName,
        index: data.color.index,
        title: data.color.title,
        isCustom: data.color.isCustom,
        clrNeutral: data.color.clrNeutral,
        colorKey: data.color.key,
        categoryKey: data.categoryKey
    };
    newActiveRound.teamA.color = data.color.clrA;
    newActiveRound.teamB.color = data.color.clrB;
    activeRound.value = newActiveRound;
}

function getSelectedColorGroup(): ColorGroup {
    return perGameData[runtimeConfig.value.gameVersion].colors.find(group =>
        group.meta.name === activeRound.value.activeColor.categoryName);
}

export function getNextColor(): ColorWithCategory {
    const selectedColorGroup = getSelectedColorGroup();
    const selectedIndex = activeRound.value.activeColor.index;
    const nextColor = selectedColorGroup.colors.find(color =>
        color.index === (selectedIndex + 1 === selectedColorGroup.colors.length ? 0 : selectedIndex + 1));

    return {
        ...(swapColorsInternally.value ? swapColors(nextColor) : nextColor),
        categoryName: selectedColorGroup.meta.name,
        categoryKey: selectedColorGroup.meta.key
    };
}

export function getPreviousColor(): ColorWithCategory {
    const selectedColorGroup = getSelectedColorGroup();
    const selectedIndex = activeRound.value.activeColor.index;
    const previousColor = selectedColorGroup.colors.find(color =>
        color.index === (selectedIndex === 0 ? selectedColorGroup.colors.length - 1 : selectedIndex - 1));

    return {
        ...(swapColorsInternally.value ? swapColors(previousColor) : previousColor),
        categoryName: selectedColorGroup.meta.name,
        categoryKey: selectedColorGroup.meta.key
    };
}
