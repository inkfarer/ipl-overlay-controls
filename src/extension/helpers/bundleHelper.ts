import { existsSync, readdirSync, readFileSync } from 'fs';
import * as nodecgContext from './nodecg';
import { NodecgBundle } from '../../types/nodecgBundle';
import { GameVersion } from '../../types/enums/gameVersion';
import i18next from 'i18next';

const nodecg = nodecgContext.get();
const bundlesDir = `${process.cwd()}/bundles`;
const allGameVersions = Object.keys(GameVersion);

export const dependentBundles: Array<NodecgBundle> = readdirSync(bundlesDir, { withFileTypes: true })
    .filter(dir => dir.isDirectory() && dir.name !== nodecg.bundleName)
    .map(bundle => {
        const bundlePackagePath = `${bundlesDir}/${bundle.name}/package.json`;
        if (existsSync(bundlePackagePath)) {
            const parsedPackage = JSON.parse(readFileSync(bundlePackagePath, 'utf-8'));
            const gameVersions = normalizeGameVersions(parsedPackage.compatibleGameVersions);
            if (gameVersions.length <= 0) {
                nodecg.log.warn(i18next.t('bundleHelper.gameVersionParsingFailedWarning', { bundleName: bundle.name }));
            }

            if ('compatibleDashboardVersion' in parsedPackage) {
                return ({
                    name: bundle.name,
                    compatibleDashboardVersion: parsedPackage.compatibleDashboardVersion,
                    compatibleGameVersions: gameVersions ?? []
                });
            }
        }

        return null;
    })
    .filter(Boolean);

function normalizeGameVersions(versions?: Array<string> | null): Array<string> {
    if (!Array.isArray(versions) || versions.length <= 0) {
        return [ GameVersion.SPLATOON_2 ];
    }

    return versions.filter(version => allGameVersions.includes(version));
}
