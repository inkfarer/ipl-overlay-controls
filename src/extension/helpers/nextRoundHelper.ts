import isEmpty from 'lodash/isEmpty';
import { getTeam } from './tournamentDataHelper';
import * as nodecgContext from './nodecg';
import { NextRound, RoundStore, TournamentData } from 'schemas';
import i18next from 'i18next';

const nodecg = nodecgContext.get();

const nextRound = nodecg.Replicant<NextRound>('nextRound');
const tournamentData = nodecg.Replicant<TournamentData>('tournamentData');
const roundStore = nodecg.Replicant<RoundStore>('roundStore');

export function setNextRoundGames(roundId: string): void {
    const round = roundStore.value[roundId];
    if (isEmpty(round)) {
        throw new Error(i18next.t('nextRoundHelper.roundNotFound', { roundId }));
    }

    nextRound.value.round = {
        id: roundId,
        name: round.meta.name,
        type: round.meta.type
    };
    nextRound.value.games = round.games.map(game => ({ stage: game.stage, mode: game.mode }));
}

export function setNextRoundTeams(teamAId: string, teamBId: string): void {
    const teamA = getTeam(teamAId, tournamentData.value);
    const teamB = getTeam(teamBId, tournamentData.value);
    if ([teamA, teamB].filter(isEmpty).length > 0) {
        throw new Error(i18next.t('nextRoundHelper.teamNotFound'));
    }

    nextRound.value.teamA = teamA;
    nextRound.value.teamB = teamB;
}

