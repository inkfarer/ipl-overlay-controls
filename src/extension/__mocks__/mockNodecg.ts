import express from 'express';
import last from 'lodash/last';
import { Configschema } from '../../types/schemas';
import cloneDeep from 'lodash/cloneDeep';

export type ReplicantChangeHandler = (newValue?: unknown, oldValue?: unknown) => void;
type OnFunction = (event: string, handler: ReplicantChangeHandler) => void;

const defaultBundleConfig = Object.freeze({
    lastfm: {
        apiKey: 'lastfmkey123',
        secret: 'lastfmsecret456'
    },
    smashgg: {
        apiKey: 'smashggkey789'
    },
    radia: {
        url: 'radia://url',
        socketUrl: 'ws://radia.url',
        authentication: 'radia-auth-12345'
    }
});

export let replicants: {[key: string]: unknown} = {};
export const replicantChangeListeners: {[key: string]: ReplicantChangeHandler} = {};
export const requestHandlers: {[type in 'POST' | 'GET']: {[path: string]: express.RequestHandler}}
    = { POST: {}, GET: {} };
export const mockBundleConfig: Configschema = cloneDeep(defaultBundleConfig);
export const messageListeners: {[key: string]: (message?: unknown, cb?: () => void) => void} = {};
export const mockMount = jest.fn();
export const mockNodecgLog = {
    warn: jest.fn(),
    info: jest.fn(),
    error: jest.fn()
};
export const mockSendMessage = jest.fn();

beforeEach(() => {
    replicants = {};
    Object.assign(mockBundleConfig, cloneDeep(defaultBundleConfig));
});

require('../helpers/nodecg').set({
    Replicant(name: string) {
        const replicantValue: { value: unknown, on: OnFunction } = {
            get value() {
                return replicants[name];
            },
            set value(newValue: unknown) {
                replicants[name] = newValue;
            },
            on: (event: string, handler: ReplicantChangeHandler) => {
                replicantChangeListeners[name] = handler;
            }
        };
        return replicantValue;
    },
    listenFor: (messageName: string, handler: () => void) => {
        messageListeners[messageName] = handler as (message: unknown, cb?: () => void) => void;
    },
    Router: () => ({
        post(path: string, ...handlers: express.RequestHandler[]) {
            requestHandlers['POST'][path] = last(handlers);
        },
        get(path: string, ...handlers: express.RequestHandler[]) {
            requestHandlers['GET'][path] = last(handlers);
        }
    }),
    mount: mockMount,
    log: mockNodecgLog,
    bundleConfig: mockBundleConfig,
    bundleName: 'ipl-overlay-controls',
    sendMessage: mockSendMessage
});
