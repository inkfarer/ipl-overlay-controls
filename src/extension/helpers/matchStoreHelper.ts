import * as nodecgContext from '../helpers/nodecg';
import { generateId } from '../../helpers/generateId';
import cloneDeep from 'lodash/cloneDeep';
import { GameWinner } from '../../types/enums/gameWinner';
import {
    ActiveRound,
    MatchStore,
    RoundStore,
    RuntimeConfig,
    SwapColorsInternally,
    TournamentData
} from '../../types/schemas';
import { setActiveRoundGames, setActiveRoundTeams } from './activeRoundHelper';
import { perGameData } from '../../helpers/gameData/gameData';

const nodecg = nodecgContext.get();
const matchStore = nodecg.Replicant<MatchStore>('matchStore');
const roundStore = nodecg.Replicant<RoundStore>('roundStore');
const tournamentData = nodecg.Replicant<TournamentData>('tournamentData');
const activeRound = nodecg.Replicant<ActiveRound>('activeRound');
const runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');
const swapColorsInternally = nodecg.Replicant<SwapColorsInternally>('swapColorsInternally');

export function resetMatchStore(resetColors = false): void {
    const firstRoundId = Object.keys(roundStore.value)[0];
    const firstRound = roundStore.value[firstRoundId];
    const matchId = generateId();

    const firstTeam = cloneDeep(tournamentData.value.teams[0]);
    const secondTeam = cloneDeep(tournamentData.value.teams[1]);

    matchStore.value = {
        [matchId]: {
            meta: {
                name: firstRound.meta.name,
                isCompleted: false,
                type: firstRound.meta.type
            },
            teamA: {
                ...firstTeam,
                score: 0
            },
            teamB: {
                ...secondTeam,
                score: 0
            },
            games: firstRound.games.map(game => ({ ...game, winner: GameWinner.NO_WINNER }))
        }
    };

    setActiveRoundToFirstMatch(resetColors);
}

export function setActiveRoundToFirstMatch(resetColors = false): void {
    const firstMatchId = Object.keys(matchStore.value)[0];
    const firstMatch = matchStore.value[firstMatchId];

    const newActiveRound = cloneDeep(activeRound.value);
    setActiveRoundGames(newActiveRound, firstMatchId);
    setActiveRoundTeams(newActiveRound, firstMatch.teamA.id, firstMatch.teamB.id);
    if (resetColors) {
        const gameData = perGameData[runtimeConfig.value.gameVersion];
        const firstCategory = gameData.colors[0];
        const firstColor = firstCategory.colors[0];
        newActiveRound.teamA.color = swapColorsInternally.value ? firstColor.clrB : firstColor.clrA;
        newActiveRound.teamB.color = swapColorsInternally.value ? firstColor.clrA : firstColor.clrB;
        newActiveRound.activeColor = {
            index: firstColor.index,
            title: firstColor.title,
            categoryName: firstCategory.meta.name,
            isCustom: firstColor.isCustom,
            clrNeutral: firstColor.clrNeutral
        };
    }

    activeRound.value = newActiveRound;
}
