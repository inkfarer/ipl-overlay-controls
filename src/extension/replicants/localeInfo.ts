import * as nodecgContext from '../helpers/nodecg';
import { LocaleInfo, RuntimeConfig } from '../../types/schemas';
import { Locale } from '../../types/enums/Locale';
import { GameVersion } from '../../types/enums/gameVersion';
import { perGameData } from '../../helpers/gameData/gameData';

const nodecg = nodecgContext.get();

const localeInfo = nodecg.Replicant<LocaleInfo>('localeInfo');

export function initLocaleInfoIfNeeded(runtimeConfig: RuntimeConfig): void {
    if (Object.keys(localeInfo.value.modes).length <= 0 && Object.keys(localeInfo.value.stages).length <= 0) {
        updateLocaleInfo(runtimeConfig.locale as Locale, runtimeConfig.gameVersion as GameVersion);
    }
}

export function updateLocaleInfo(locale: Locale, gameVersion: GameVersion): void {
    const gameData = perGameData[gameVersion];

    localeInfo.value = {
        stages: gameData.stages[locale],
        modes: gameData.modes[locale]
    };
}
