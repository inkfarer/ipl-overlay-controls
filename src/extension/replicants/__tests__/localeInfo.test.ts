import { replicants } from '../../__mocks__/mockNodecg';
import { initLocaleInfoIfNeeded, updateLocaleInfo } from '../localeInfo';
import { Locale } from '../../../types/enums/Locale';
import { GameVersion } from '../../../types/enums/gameVersion';
import { cartesian } from '../../../helpers/ArrayHelper';

describe('localeInfo', () => {
    describe('initLocaleInfoIfNeeded', () => {
        it('does nothing if locale info is present', () => {
            const originalLocaleInfo = {
                modes: {
                    'Turf War': 'TURF'
                },
                stages: {
                    'Ancho-V Games': 'ANCHOV'
                }
            };
            replicants.localeInfo = originalLocaleInfo;

            initLocaleInfoIfNeeded({ locale: Locale.DE, gameVersion: GameVersion.SPLATOON_2 });

            expect(replicants.localeInfo).toEqual(originalLocaleInfo);
        });

        it('updates locale info if no locale info is present', () => {
            replicants.localeInfo = {
                modes: { },
                stages: { }
            };

            initLocaleInfoIfNeeded({ locale: Locale.DE, gameVersion: GameVersion.SPLATOON_2 });

            expect(replicants.localeInfo).toEqual({
                modes: {
                    'Clam Blitz': 'Muschelchaos',
                    Rainmaker: 'Operation Goldfisch',
                    'Splat Zones': 'Herrschaft',
                    'Tower Control': 'Turmkommando',
                    'Turf War': 'Revierkampf',
                    'Unknown Mode': 'Unbekannte Kampfart'
                },
                stages: {
                    'Ancho-V Games': 'Anchobit Games HQ',
                    'Arowana Mall': 'Arowana Center',
                    'Blackbelly Skatepark': 'Punkasius-Skatepark',
                    'Camp Triggerfish': 'Camp Schützenfisch',
                    Counterpick: 'Counterpick',
                    'Goby Arena': 'Backfisch-Stadion',
                    'Humpback Pump Track': 'Buckelwal-Piste',
                    'Inkblot Art Academy': 'Perlmutt-Akademie',
                    'Kelp Dome': 'Tümmlerkuppel',
                    MakoMart: 'Cetacea-Markt',
                    'Manta Maria': 'Manta Maria',
                    'Moray Towers': 'Muränentürme',
                    'Musselforge Fitness': 'Molluskelbude',
                    'New Albacore Hotel': 'Hotel Neothun',
                    'Piranha Pit': 'Steinköhler-Grube',
                    'Port Mackerel': 'Heilbutt-Hafen',
                    'Shellendorf Institute': 'Abyssal-Museum',
                    'Shifty Station': 'Wandelzone',
                    'Skipper Pavilion': 'Grundel-Pavillon',
                    'Snapper Canal': 'Grätenkanal',
                    'Starfish Mainstage': 'Seeigel-Rockbühne',
                    'Sturgeon Shipyard': 'Störwerft',
                    'The Reef': 'Korallenviertel',
                    'Unknown Stage': 'Unbekannte Arena',
                    'Wahoo World': 'Flunder-Funpark',
                    'Walleye Warehouse': 'Kofferfisch-Lager',
                }
            });
        });
    });

    describe('updateLocaleInfo', () => {
        it.each(cartesian(Object.values(GameVersion), Object.values(Locale)))(
            'generates expected locale info when game version is %s and locale is %s',
            (gameVersion, locale) => {
                updateLocaleInfo(locale, gameVersion);

                expect(replicants.localeInfo).toMatchSnapshot();
            });
    });
});
