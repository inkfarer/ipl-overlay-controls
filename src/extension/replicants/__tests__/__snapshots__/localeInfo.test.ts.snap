// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`localeInfo updateLocaleInfo generates expected locale info when game version is SPLATOON_2 and locale is DE 1`] = `
Object {
  "modes": Object {
    "Clam Blitz": "Muschelchaos",
    "Rainmaker": "Operation Goldfisch",
    "Splat Zones": "Herrschaft",
    "Tower Control": "Turmkommando",
    "Turf War": "Revierkampf",
    "Unknown Mode": "Unbekannte Kampfart",
  },
  "stages": Object {
    "Ancho-V Games": "Anchobit Games HQ",
    "Arowana Mall": "Arowana Center",
    "Blackbelly Skatepark": "Punkasius-Skatepark",
    "Camp Triggerfish": "Camp Schützenfisch",
    "Counterpick": "Counterpick",
    "Goby Arena": "Backfisch-Stadion",
    "Humpback Pump Track": "Buckelwal-Piste",
    "Inkblot Art Academy": "Perlmutt-Akademie",
    "Kelp Dome": "Tümmlerkuppel",
    "MakoMart": "Cetacea-Markt",
    "Manta Maria": "Manta Maria",
    "Moray Towers": "Muränentürme",
    "Musselforge Fitness": "Molluskelbude",
    "New Albacore Hotel": "Hotel Neothun",
    "Piranha Pit": "Steinköhler-Grube",
    "Port Mackerel": "Heilbutt-Hafen",
    "Shellendorf Institute": "Abyssal-Museum",
    "Shifty Station": "Wandelzone",
    "Skipper Pavilion": "Grundel-Pavillon",
    "Snapper Canal": "Grätenkanal",
    "Starfish Mainstage": "Seeigel-Rockbühne",
    "Sturgeon Shipyard": "Störwerft",
    "The Reef": "Korallenviertel",
    "Unknown Stage": "Unbekannte Arena",
    "Wahoo World": "Flunder-Funpark",
    "Walleye Warehouse": "Kofferfisch-Lager",
  },
}
`;

exports[`localeInfo updateLocaleInfo generates expected locale info when game version is SPLATOON_2 and locale is EN 1`] = `
Object {
  "modes": Object {
    "Clam Blitz": "Clam Blitz",
    "Rainmaker": "Rainmaker",
    "Splat Zones": "Splat Zones",
    "Tower Control": "Tower Control",
    "Turf War": "Turf War",
    "Unknown Mode": "Unknown Mode",
  },
  "stages": Object {
    "Ancho-V Games": "Ancho-V Games",
    "Arowana Mall": "Arowana Mall",
    "Blackbelly Skatepark": "Blackbelly Skatepark",
    "Camp Triggerfish": "Camp Triggerfish",
    "Counterpick": "Counterpick",
    "Goby Arena": "Goby Arena",
    "Humpback Pump Track": "Humpback Pump Track",
    "Inkblot Art Academy": "Inkblot Art Academy",
    "Kelp Dome": "Kelp Dome",
    "MakoMart": "MakoMart",
    "Manta Maria": "Manta Maria",
    "Moray Towers": "Moray Towers",
    "Musselforge Fitness": "Musselforge Fitness",
    "New Albacore Hotel": "New Albacore Hotel",
    "Piranha Pit": "Piranha Pit",
    "Port Mackerel": "Port Mackerel",
    "Shellendorf Institute": "Shellendorf Institute",
    "Shifty Station": "Shifty Station",
    "Skipper Pavilion": "Skipper Pavilion",
    "Snapper Canal": "Snapper Canal",
    "Starfish Mainstage": "Starfish Mainstage",
    "Sturgeon Shipyard": "Sturgeon Shipyard",
    "The Reef": "The Reef",
    "Unknown Stage": "Unknown Stage",
    "Wahoo World": "Wahoo World",
    "Walleye Warehouse": "Walleye Warehouse",
  },
}
`;

exports[`localeInfo updateLocaleInfo generates expected locale info when game version is SPLATOON_3 and locale is DE 1`] = `
Object {
  "modes": Object {
    "Clam Blitz": "Muschelchaos",
    "Rainmaker": "Operation Goldfisch",
    "Splat Zones": "Herrschaft",
    "Tower Control": "Turmkommando",
    "Turf War": "Revierkampf",
    "Unknown Mode": "Unbekannte Kampfart",
  },
  "stages": Object {
    "Brinewater Springs": "Kusaya-Quellen",
    "Counterpick": "Counterpick",
    "Eeltail Alley": "Streifenaal-Straße",
    "Flounder Heights": "Schollensiedlung",
    "Hagglefish Market": "Schnapperchen-Basar",
    "Hammerhead Bridge": "Makrelenbrücke",
    "Inkblot Art Academy": "Perlmutt-Akademie",
    "Mahi-Mahi Resort": "Mahi-Mahi-Resort",
    "MakoMart": "Cetacea-Markt",
    "Manta Maria": "Manta Maria",
    "Mincemeat Metalworks": "Aalstahl-Metallwerk",
    "Museum d'Alfonsino": "Pinakoithek",
    "Scorch Gorge": "Sengkluft",
    "Sturgeon Shipyard": "Störwerft",
    "Um'ami Ruins": "Um'ami-Ruinen",
    "Undertow Spillway": "Schwertmuschel-Reservoir",
    "Unknown Stage": "Unbekannte Arena",
    "Wahoo World": "Flunder-Funpark",
  },
}
`;

exports[`localeInfo updateLocaleInfo generates expected locale info when game version is SPLATOON_3 and locale is EN 1`] = `
Object {
  "modes": Object {
    "Clam Blitz": "Clam Blitz",
    "Rainmaker": "Rainmaker",
    "Splat Zones": "Splat Zones",
    "Tower Control": "Tower Control",
    "Turf War": "Turf War",
    "Unknown Mode": "Unknown Mode",
  },
  "stages": Object {
    "Brinewater Springs": "Brinewater Springs",
    "Counterpick": "Counterpick",
    "Eeltail Alley": "Eeltail Alley",
    "Flounder Heights": "Flounder Heights",
    "Hagglefish Market": "Hagglefish Market",
    "Hammerhead Bridge": "Hammerhead Bridge",
    "Inkblot Art Academy": "Inkblot Art Academy",
    "Mahi-Mahi Resort": "Mahi-Mahi Resort",
    "MakoMart": "MakoMart",
    "Manta Maria": "Manta Maria",
    "Mincemeat Metalworks": "Mincemeat Metalworks",
    "Museum d'Alfonsino": "Museum d'Alfonsino",
    "Scorch Gorge": "Scorch Gorge",
    "Sturgeon Shipyard": "Sturgeon Shipyard",
    "Um'ami Ruins": "Um'ami Ruins",
    "Undertow Spillway": "Undertow Spillway",
    "Unknown Stage": "Unknown Stage",
    "Wahoo World": "Wahoo World",
  },
}
`;
